platform/metrics/metrics-servlet-filter/src/main/java/org/codice/ddf/metrics/servlet/ServletMetrics.java

```diff
@@ -0,0 +1,151 @@
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.ddf.metrics.servlet;
+
+import io.micrometer.core.instrument.DistributionSummary;
+import io.micrometer.core.instrument.Metrics;
+import java.io.IOException;
+import javax.servlet.AsyncEvent;
+import javax.servlet.AsyncListener;
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ServletMetrics implements Filter {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(ServletMetrics.class);
+
+  private static final String METRICS_PREFIX = "ddf.platform.http";
+
+  private static final String HISTOGRAM_NAME = "latency";
+
+  @Override
+  public void init(FilterConfig filterConfig) throws ServletException {
+    LOGGER.debug("Adding metrics security filter.");
+  }
+
+  @Override
+  public void doFilter(
+      ServletRequest servletRequest, ServletResponse servletResponse, FilterChain chain)
+      throws IOException, ServletException {
+    boolean hadException = false;
+    long startTime = System.currentTimeMillis();
+    try {
+      chain.doFilter(servletRequest, servletResponse);
+    } catch (Exception ex) {
+      hadException = true;
+      throw ex;
+    } finally {
+      HttpServletRequest request = (HttpServletRequest) servletRequest;
+      HttpServletResponse response = (HttpServletResponse) servletResponse;
+
+      if (!hadException && request.isAsyncStarted()) {
+        request.getAsyncContext().addListener(new AsyncResponseListener(startTime));
+      } else {
+        record(request, response, startTime, hadException, false);
+      }
+    }
+  }
+
+  private static void record(
+      HttpServletRequest request,
+      HttpServletResponse response,
+      long startTime,
+      boolean hadException,
+      boolean hadTimeout) {
+    long endTime = System.currentTimeMillis();
+    long latency = endTime - startTime;
+
+    DistributionSummary.builder(METRICS_PREFIX + "." + HISTOGRAM_NAME)
+        .baseUnit("milliseconds")
+        .tags(
+            "method",
+            request.getMethod(),
+            "status",
+            getStatusCode(response, hadException, hadTimeout))
+        .publishPercentiles(0.5, 0.95)
+        .register(Metrics.globalRegistry)
+        .record(latency);
+  }
+
+  private static String getStatusCode(
+      HttpServletResponse response, boolean hadException, boolean hadTimeout) {
+    String result = String.valueOf(response.getStatus());
+    if (hadException && response.getStatus() != 500) {
```

Comment:
❓ Looks unlikely, but could there be an exception AND a timeout? If so the status code will get 'confused'

Reply:
The API was not clear on that.
https://docs.oracle.com/javaee/7/api/javax/servlet/AsyncListener.html

Currently it does not appear that would happen with Jetty.
https://github.com/eclipse/jetty.project/blob/1d734e1daff393968b231626cf223e76f605b0c3/jetty-servlets/src/test/java/org/eclipse/jetty/servlets/AsyncTimeoutCompleteWrite.java#L38-L42

It is ordered so that less specific case (`hadException`) will log but the more specific case (`hadTimeout`) will log and return for metrics tagging.